File: frontend/src/graphql/queries/review.ts
import { gql } from '@apollo/client';

export const GET_COURSE_REVIEWS = gql`
    query GetCourseReviews($courseId: UUID!) {
        getReviewsByCourse(courseId: $courseId) {
            reviewId
            rating
            comment
            createdAt
            user {
                userId
                fullName
                avatarUrl
            }
        }
        getCourseAverageRating(courseId: $courseId)
    }
`;

export const GET_MY_REVIEW_FOR_COURSE = gql`
    query GetMyReviewForCourse($courseId: UUID!) {
        myReviewForCourse(courseId: $courseId) {
            reviewId
            rating
            comment
            createdAt
            user {
                userId
                fullName
                avatarUrl
            }
        }
    }
`;

File: backend/src/main/java/com/seikyuuressha/lms/entity/Review.java
package com.seikyuuressha.lms.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "Reviews", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"userId", "courseId"})
})
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Review {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    UUID reviewId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "courseId", nullable = false)
    Course course;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "userId", nullable = false)
    Users user;

    @Column(nullable = false)
    Integer rating; // 1-5 stars

    @Column(columnDefinition = "TEXT")
    String comment;

    @Column(nullable = false)
    Boolean isActive;

    @Column(nullable = false)
    LocalDateTime createdAt;

    LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (reviewId == null) {
            reviewId = UUID.randomUUID();
        }
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (isActive == null) {
            isActive = true;
        }
        // Validate rating
        if (rating != null && (rating < 1 || rating > 5)) {
            throw new IllegalArgumentException("Rating must be between 1 and 5");
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
